### PR TITLE
Addsession

### PR DESCRIPTION
**Pull request by gabidoye**

### CODE REVIEW

File: gpa.py [hunk 1]
```diff
@@ -68,14 +58,14 @@ def compute_student_gpa(units, scores):
...
    total_grade_point = sum([a*b for a,b in zip(units,points)])# 11 [a,a,e][5,5,1][3,4,2]
-   gpa= round((total_grade_point / total_unit), 2)
+   gpa = round((total_grade_point / total_unit), 2)
 
    return total_unit,total_grade_point,gpa
 
-def get_score_data(term,course_records):
+def get_score_data(session,term,course_records):
    """
    This function takes the term and student record and extracts the scores and the units for the semester.
```
Comment (manuelibidun) [inline gpa.py]: update the docstring?

File: gpa.py [hunk 2]
```diff
@@ -103,13 +93,43 @@ def get_semesters(course_records):
...
+   dict
+   return
+   -------
+   array of sessions
+   """
+   sessions = set()
+   for d in (course_records['courses']):
+      for key, value in d.items():
+         if key == 'session':
+               sessions.add(value)
```
Comment (manuelibidun) [inline gpa.py]:
```suggestion
sessions = set([course['session'] for course in course_records['courses']])
```

File: gpa.py [hunk 3]
```diff
@@ -103,13 +93,43 @@ def get_semesters(course_records):
    -------
    array of semesters
    """
-   semesters=set()
+   semesters = set()
    for d in (course_records['courses']):
       for key, value in d.items():
-         if key=='term':
+         if key == 'term':
                semesters.add(value)
```
Comment (manuelibidun) [inline gpa.py]: Replace this entire block with 
`semesters = set([course['term'] for course in course_records['courses']])`

File: gpa.py [hunk 4]
```diff
@@ -121,24 +141,44 @@ def compute_results(student_records):
...
+      session_total_grade_points.append(total_grade_points)
+      
+      """
+      final year computaion of cgpa with values of each session
+      """
+      final_year_cgpa = round(sum(calculate_sum(session_total_grade_points)) / sum(calculate_sum(session_total_units)),2)
+      final_year_results['name'] = student_records['name']
+      final_year_results['id'] = student_records['id']
+      final_year_results[f"{session}_session"] = results
+      final_year_results['final_year_cgpa'] = final_year_cgpa
```
Comment (manuelibidun) [inline gpa.py]: Why is this nested in a for loop?

Comment (manuelibidun) [reply gpa.py]: And by the way, I'm not sure the current solution will be able to report details such as: 
Year: 2019, Level: 100L
Semester: Fall
     total unit, GPA, CGPA
Semester: Spring
     total unit, GPA, CGPA
Year: 2020, Level: 200L
Semester: Fall
     total unit, GPA, CGPA
Semester: Spring
     total unit, GPA, CGPA
.....
.....
Year: 2022, Level: 400L
Semester: Fall
     total unit, GPA, CGPA
Semester: Spring
     total unit, GPA, CGPA
.....
total units, CPGA

File: gpa.py [hunk 5]
```diff
@@ -147,11 +187,11 @@ def compute_results(student_records):
...
-         {"title": "English", "unit": 2, "code": "ENG101", "score": 60, "term": "Fall"},
-         {"title": "Chemistry I", "unit": 4, "code": "CHE101", "score": 70, "term": "Fall"},
-         {"title": "Maths", "unit": 3, "code": "MTH101", "score": 80, "term": "Spring"},
-         {"title": "Chemistry II", "unit": 4, "code": "CHE102", "score": 91, "term": "Spring"}, 
-         {"title": "History", "unit": 2, "code": "HIS102", "score": 40, "term": "Spring"}
+         {"title": "English", "unit": 2, "code": "ENG101", "score": 60, "term": "Fall", "session":2020},
+         {"title": "Chemistry I", "unit": 4, "code": "CHE101", "score": 70, "term": "Fall", "session":2021},
+         {"title": "Maths", "unit": 3, "code": "MTH101", "score": 80, "term": "Spring", "session":2020},
+         {"title": "Chemistry II", "unit": 4, "code": "CHE102", "score": 91, "term": "Spring", "session":2021}, 
+         {"title": "History", "unit": 2, "code": "HIS102", "score": 40, "term": "Spring", "session":2020}
```
Comment (manuelibidun) [inline gpa.py]: I'd recommend to make session a string. So you should be using "2021" rather than a plain integer.

**Review by manuelibidun — CHANGES_REQUESTED** [review]

File: gpa.py (hunk 5 above)
Comment (gabidoye) [reply gpa.py]: Resolved

File: gpa.py (hunk 2 above)
Comment (gabidoye) [reply gpa.py]: Done

File: gpa.py (hunk 3 above)
Comment (gabidoye) [reply gpa.py]: DOne

File: gpa.py (hunk 1 above)
Comment (gabidoye) [reply gpa.py]: Done

File: gpa.py (hunk 4 above)
Comment (gabidoye) [reply gpa.py]: I moved the final computation outside the nested loop.
Refactored the code. Used a different data structure(Tuple) as the dictionary gave some issues around duplicate keys during the loop.

**Review by manuelibidun — APPROVED** [review]

**Pull request merged**